File: AlgorithmReview/Programmers/Lv2/피로도.swift
//
//  피로도.swift
//  AlgorithmReview
//
//  Created by 박소윤 on 2023/05/09.
//

import Foundation

func solution(_ k:Int, _ dungeons:[[Int]]) -> Int {

    var orders = [[Int]]()
    var visit = [Bool](repeating: false, count: dungeons.count)
    func generate(_ order:[Int]) {
        
        if order.count == dungeons.count {
            orders.append(order)
            return
        }
        
        for i in 0..<dungeons.count {
            if !visit[i] {
                visit[i] = true
                generate(order + [i])
                visit[i] = false
            }
        }
    }
    
    generate([])
    var result = 0
    for order in orders {
        var remain = k
        var complete = 0
        for d in order {
            if remain >= dungeons[d][0] {
                remain -= dungeons[d][1]
                complete += 1
            }
            else {
                break
            }
        }
        result = max(result, complete)
    }
    
    return result
}

File: Python/피로도.py
def solution(k, dungeons):
    
    orders = []
    visit = [False] * len(dungeons)
    
    def generate(d):
        
        if len(d) == len(dungeons):
            orders.append(d)
            return
        
        for i in range(0, len(dungeons)):
            if visit[i] == False:
                visit[i] = True
                generate(d + [i])
                visit[i] = False
            
    generate([])
    
    answer = 0
    for order in orders:
        remain = k
        complete = 0
        for i,stage in enumerate(order):
            if remain >= dungeons[stage][0]:
                remain -= dungeons[stage][1]
                complete += 1
            else:
                break
        answer = max(answer, complete)
                
    return answer
